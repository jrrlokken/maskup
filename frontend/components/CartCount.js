import { TransitionGroup, CSSTransition } from 'react-transition-group';
import styled from 'styled-components';

const Badge = styled.div`
  background-color: var(--black);
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(2) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit {
    transform: scale(2) rotateX(0.5turn);
  }
`;

const CartCount = ({ count }) => (
  <AnimationStyles>
   <TransitionGroup>
     <CSSTransition
      className='count'
      classNames='count'
      key={count}
      timeout={{ enter: 400, exit: 400}}
      unmountOnExit
      >
      <Badge>{count}</Badge>
     </CSSTransition>
   </TransitionGroup>
  </AnimationStyles>
);

export default CartCount;