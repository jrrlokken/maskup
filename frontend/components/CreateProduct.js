import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { useRouter } from 'next/router';

import { ALL_PRODUCTS_QUERY } from './Products';
import useForm from '../lib/useForm';
import DisplayError from './ErrorMessage';
import Form from './styles/Form';
import StyledButton from './styles/StyledButton';

const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      name
      price
      description
    }
  }
`;

const CreateProduct = () => {
  const { inputs, handleChange, clearForm } = useForm({
    name: '',
    description: '',
    image: '',
    price: 0,
  });

  const [createProduct, { data, error, loading }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      variables: inputs,
      refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
    }
  );

  const router = useRouter();

  async function handleSubmit(event) {
    event.preventDefault();
    const res = await createProduct();
    console.log(res);
    clearForm();
    router.push(`/product/${res.data.createProduct.id}`);
  }

  return (
    <Form onSubmit={handleSubmit}>
      <DisplayError error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor='image'>
          Image
          <input
            required
            type='file' 
            id='image' 
            name='image'
            onChange={handleChange}
          />
        </label>
        <label htmlFor='name'>
          Name
          <input
            required
            type='text' 
            id='name' 
            name='name' 
            placeholder='Name'
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor='description'>
          Description
          <textarea
            required
            id='description' 
            name='description' 
            placeholder='Description'
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <label htmlFor='price'>
          Price
          <input
            required
            type='number' 
            id='price' 
            name='price' 
            placeholder='Price'
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <StyledButton type='submit'>+ Add Product</StyledButton>
      </fieldset>
    </Form>
  );
}

export default CreateProduct;