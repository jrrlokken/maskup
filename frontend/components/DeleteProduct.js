import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { useRouter } from 'next/router';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

const update = (cache, payload) => {
  cache.evict(cache.identify(payload.data.deleteProduct));
}

const DeleteProduct = ({ id, children }) => {
  const [deleteProduct, { error, loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
    update,
  });
  const router = useRouter();

  const handleClick = () => {
    if (confirm('Are you sure?')) {
      deleteProduct()
        .catch(error => {
          alert(error.message);
        });
    }
    router.push('/products');
  }

  return (
    <button
      type='button'
      disabled={loading}
      onClick={handleClick}
    > 
      {children}
       🗑️
    </button>
  );
}

export default DeleteProduct;
export { DELETE_PRODUCT_MUTATION };
