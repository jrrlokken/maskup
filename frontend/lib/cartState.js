import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

const CartStateProvider = ({ children }) => {
  const [cartOpen, setCartOpen] = useState(false);
  const toggleCart = () => {
    setCartOpen(!cartOpen);
  }
  const closeCart = () => {
    setCartOpen(false);
  }
  const openCart = () => {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{
        cartOpen,
        setCartOpen,
        toggleCart,
        closeCart,
        openCart
      }}
    >
      {children}
    </LocalStateProvider>
  );
}

const useCart = () => {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };