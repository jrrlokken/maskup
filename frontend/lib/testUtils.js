import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const type = (wrapper, name, value) => {
  wrapper.find(`input[name='${name}']`).simulate('change', {
    target: { name, value }
  });
};

const fakeProduct = ({
  // __typename: 'Item',
  id: 'abc123',
  name: 'Test Product',
  description: 'This is a test product',
  price: 1499,
  photo: {
    altText: 'dog',
    image: {
      publicUrlTransformed: 'dog.jpg',
    },
  },
  user: null,
});

const fakeUser = overrides => ({
  // __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  // __typename: 'OrderItem',
  id: casual.uuid,
  photo: {
    image: {
      publicUrlTransformed: `${casual.word}.jpg`
    }
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  // __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = overrides => ({
  // __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeProduct,
  user: fakeUser(),
  ...overrides,
});


// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  type,
  fakeProduct,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  LocalStorageMock,
};