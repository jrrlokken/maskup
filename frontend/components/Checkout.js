import { useState } from 'react';
import { loadStripe } from '@stripe/stripe-js';
import {CardElement, Elements, useStripe, useElements } from '@stripe/react-stripe-js';
import { useMutation } from '@apollo/client';
import NProgress from 'nprogress';
import gql from 'graphql-tag';
import { useRouter } from 'next/dist/client/router';
import styled from 'styled-components';

import { CURRENT_USER_QUERY } from './User';
import { useCart } from '../lib/cartState';
import StyledButton from './styles/StyledButton';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0,0,0,0.04);
  border: 1px solid rgba(0,0,0,0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const { closeCart } = useCart();
  const [checkout, { error: graphQLError }] = useMutation(CREATE_ORDER_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  async function handleSubmit(event) {
    event.preventDefault();
    setLoading(true);
    NProgress.start();

    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });

    if (error) {
      setError(error);
      NProgress.done();
      return;
    }

    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });

    router.push({
      pathname: '/order/[id]',
      query: { id: order.data.checkout.id },
    });

    closeCart();
    setLoading(false);
    NProgress.done();
  };

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <p style={{ fontSize: 12 }}>{error.message}</p>}
      {graphQLError && <p style={{ fontSize: 12 }}>{graphQLError.message}</p>}
      <CardElement />
      <StyledButton style={{ fontWeight: 900 }}>{loading ? 'Processing...' : 'Checkout'}</StyledButton>
    </CheckoutFormStyles>
  );
}

const Checkout = () => {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}

export { Checkout };