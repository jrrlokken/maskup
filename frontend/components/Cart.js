import styled from 'styled-components';
import { useCart } from '../lib/cartState';
import { useUser } from './User';
import CartStyles from './styles/CartStyles';
import Heading from './styles/Heading';
import CloseButton from './styles/CloseButton';
import { Checkout } from './Checkout';
import RemoveFromCart from './RemoveFromCart';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = ({ cartItem }) => {
  const { product } = cartItem;
  if (!product) return null;

  return (
    <CartItemStyles>
      <img
        width='100'
        src={product.photo.image.publicUrlTransformed}
        alt={product.name}
      />
      <div>
        <h3>{product.name}</h3>
        <p>
          {formatMoney(product.price * cartItem.quantity)}-
          <em>
            {cartItem.quantity} &times; {formatMoney(product.price)} each
          </em>
        </p>
      </div>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  );
}
const Cart = () => {
  const me = useUser();
  const { cartOpen, closeCart } = useCart();
  if (!me) return null;

  return (
    <CartStyles open={cartOpen}>
      <header>
        <Heading>{me.name}'s Cart</Heading>
        <CloseButton onClick={closeCart}>&times;</CloseButton>
      </header>
      <ul>
        {me.cart.map(cartItem => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
        <Checkout />
      </footer>
    </CartStyles>
  );
}

export default Cart;