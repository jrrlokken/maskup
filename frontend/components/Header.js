import Link from 'next/link';
import styled from 'styled-components';

import Cart from './Cart';
import Nav from './Nav';
import Search from './Search';

const Logo = styled.h1`
  font-size: 4rem;
  margin-left: 2rem;
  position: relative;
  z-index: 2;
  /* background-color: var(--blue); */
  a {
    color: white;
    text-decoration: none;
    text-transform: uppercase;
    padding: 0.5rem 1rem;
  }
  @media (max-width: 1300px) {
    margin: 2rem;
    text-align: center;
  }
`;

const StyledHeader = styled.header`
  position: sticky;
  top: 0;
  z-index: 10;
  background-color: var(--blue);
  .bar {
    border-bottom: 10px solid var(--black);
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
    @media (max-width: 1300px) {
      grid-template-columns: 1fr;
      justify-content: center;
    }
  }
  
  .sub-bar {
    display: grid;
    grid-template-columns: 1fr auto;
    border-bottom: 1px solid var(--black);
  }
`;

const Header = () => (
  <StyledHeader>
    <div className="bar">
      <Logo><Link href='/'><a id='logo'>Mask Up!</a></Link></Logo>
      <Nav />
    </div>
    <div className="sub-bar">
      <Search />
    </div>
    <Cart />
  </StyledHeader>
);

export default Header;