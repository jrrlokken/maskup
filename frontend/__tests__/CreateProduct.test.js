import { configure, mount } from 'enzyme';
import Adapter from '@hteker/enzyme-adapter-react-17';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/client/testing';
import { act } from 'react-dom/test-utils';
import wait from 'waait';

import CreateProduct, { CREATE_PRODUCT_MUTATION } from '../components/CreateProduct';
import { CURRENT_USER_QUERY } from '../components/User';
import { CartStateProvider } from '../lib/cartState';
import { fakeUser } from '../lib/testUtils';

configure({ adapter: new Adapter() });



describe('<CreateProduct/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
  });
});
