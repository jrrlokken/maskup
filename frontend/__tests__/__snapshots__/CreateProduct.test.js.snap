// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct/> renders and matches snapshot 1`] = `
<form
  className="Form-c7byzo-0 csfyM"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="image"
    >
      Image
      <input
        id="image"
        name="image"
        onChange={[Function]}
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        id="name"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <StyledButton
      type="submit"
    >
      <button
        className="StyledButton-sc-1vgaxf4-0 kkhzEF"
        type="submit"
      >
        + Add Product
      </button>
    </StyledButton>
  </fieldset>
</form>
`;
